packages/ember-data/lib/system/record-array-manager.js

```diff
@@ -121,11 +122,28 @@ export default Ember.Object.extend({
   */
   updateFilterRecordArray: function(array, filter, typeClass, record) {
     var shouldBeInArray = filter(record.getRecord());
-    var recordArrays = this.recordArraysForRecord(record);
+    var key = guidFor(record);
+    if (shouldBeInArray && shouldBeInArray.then) {
+      array.lastFilterPromises[key] = shouldBeInArray;
+      shouldBeInArray.then((result) => {
+        // fighting race conditions here
+        if (shouldBeInArray !== array.lastFilterPromises[key]) {
+          return;
+        }
+
+        delete array.lastFilterPromises[key];
+        this._updateFilterRecordArray(array, result, typeClass, record);
+      });
+    } else {
+      this._updateFilterRecordArray(array, shouldBeInArray, typeClass, record);
+    }
+  },
 
+  _updateFilterRecordArray: function(array, shouldBeInArray, typeClass, record) {
+    var recordArrays = this.recordArraysForRecord(record);
     if (shouldBeInArray) {
       this._addRecordToRecordArray(array, record);
-    } else {
+    } else if (!shouldBeInArray) {
```

Comment:
I think this is a left-over from rebasing. The `else if (!shouldBeInArray)` has been removed as part of cleaning up in [#3167](https://github.com/pangratz/data/blob/b9044fd978c2ff4fc6e6acda1800c57e59351454/packages/ember-data/lib/system/record-array-manager.js#L128)